### ISSUE TITLE
Canceling uncancelable doesn't wait

### ISSUE THREAD

**Issue by RafalSumislawski** [typelevel/cats-effect#2391]
The in CE 2 (2.5.4) `BracketLaws#bracketCaseWithPureUnitIsUncancelable` states that `F.bracketCase(fa)(f)((_, _) => F.unit) <-> F.uncancelable(fa).flatMap(f)`, and yet the following to examples behave differently.

### TestBracketCase
```
import cats.effect.concurrent.Deferred
import cats.effect.{Bracket, ExitCode, IO, IOApp}
import cats.implicits._

import scala.concurrent.duration.DurationInt

object TestBracketCase extends IOApp {

  val F = Bracket[IO, Throwable]

  override def run(args: List[String]): IO[ExitCode] = for {
    d <- Deferred[IO, Unit]
    fa = d.complete(()) >> IO.sleep(1.second) >> IO(println("fa done"))
    f = (a: Unit) => IO.unit
    fiber <- F.bracketCase(fa)(f)((_, _) => F.unit).start
    _ <- d.get
    _ <- fiber.cancel
    _ <- IO(println("cancel done"))
  } yield ExitCode.Success

}
```

Output:
```
fa done
cancel done
```

When using `bracketCase` `cancel` waits for the `fa` to complete before it itself completes.

### TestUncancelable
```
import cats.effect.concurrent.Deferred
import cats.effect.{Bracket, ExitCode, IO, IOApp}
import cats.implicits._

import scala.concurrent.duration.DurationInt

object TestUncancelable extends IOApp {

  val F = Bracket[IO, Throwable]

  override def run(args: List[String]): IO[ExitCode] = for {
    d <- Deferred[IO, Unit]
    fa = d.complete(()) >> IO.sleep(1.second) >> IO(println("fa done"))
    f = (a: Unit) => IO.unit
    fiber <- F.uncancelable(fa).flatMap(f).start
    _ <- d.get
    _ <- fiber.cancel
    _ <- IO(println("cancel done"))
  } yield ExitCode.Success

}
```
Output:
```
cancel done
```

After substituting `F.uncancelable(fa).flatMap(f)` for `F.bracketCase(fa)(f)((_, _) => F.unit)` `cancel` doesn't wait for the `fa` to complete and the application completes before `"fa done"` is printed.

**Comment by djspiewak** [comment]
Very unsurprising, I'm afraid. :-(

So this is a *very* long thread to pull on, but you can get an idea of just how long by perusing [this discussion](https://github.com/typelevel/cats-effect/discussions/1979). One of the enduring problems with CE2 is that it *doesn't* reliably wait on `cancel`, and that can be a source of some resource leaks in certain situations. CE3 corrects this entirely, though it does sometimes come at the cost of some surprising behavior (e.g. Alex's `timeout` example).

But at the end of the day… what you're describing is a known issue, and one which is probably unfixable in CE2 due to `IO`'s implementation.

**Comment by RafalSumislawski** [comment]
Does CE2 wait reliably for `Bracket`'s finaliser (#727) or `Bracket`'s acquire? Or is it more-reliable-but-still-unreliable? It seems to work in my tests, but maybe I just didn't hit the right edge-case. 

Would I run into some known issues by using an implementation of uncancelable based on the one in `Bracket`
```
  def uncancelable[A](fa: F[A]): F[A] =
    bracket(fa)(pure)(_ => unit)
```
instead of the one in IO
```
  final def uncancelable: IO[A] =
    IOCancel.uncancelable(this)
```
?
